Modules/_abc.c

```diff
@@ -177,6 +197,23 @@ _get_impl(PyObject *module, PyObject *self)
     return (_abc_data *)impl;
 }
 
+static _abc_data *
+_get_impl_optional(PyObject *module, PyObject *self)
```

Comment:
Since this can return NULL with or without an exception, I suggest changing it to:

```
static int
_get_impl_optional(_abc_data **res, PyObject *module, PyObject *self);
```

and return 0 if the attribute doesn't exist, -1 if there is an error and 1 if it succeeds. In the first two cases, set `*res = NULL`.